Add a multi-platform proposal

luxas opened the pull request:
This is a proposal for the work I've done with multiarch Kubernetes. I've worked quite a lot on it and now I've written it down so others may easily use and learn from it, while at the same time pointing out issues that we should fix in the future.

I would like to get this in before `v1.3`.

Feel free to comment on it if it's something that's unclear.

@david-mcmahon @ixdy @mikedanese @Pensu @davidopp @bgrant0607 @dlorenc @eparis @vishh @thockin @dchen1107 @lavalamp

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/26863)

<!-- Reviewable:end -->

dchen1107 commented:
@luxas thanks for your proposal, but I don't think this is required for 1.3 release. We are clearing up our 1.3 merging queue, once we are done with that, we are open for prs like this to main branch. Thanks!

luxas commented:
@dchen1107 It's okay, it would have been nice to have, but agree that it isn't so important if you're branching soon.
I would be interested in your opinion on this topic though, but understand if you have too much things to do atm and wait until after the release.

vishh commented on docs/proposals/multi-platform.md:
```diff
@@ -0,0 +1,451 @@
+
+## Abstract
+
+Kubernetes is written in Go, and Go is ported to multiple platforms.
```
nit: `Go code is portable across platforms`.

vishh commented on docs/proposals/multi-platform.md:
```diff
@@ -0,0 +1,451 @@
+When it's possible to run docker containers on a specific architecture, people also want to use Kubernetes to manage the containers.
+
+We obviously want Kubernetes to run on as many platforms as possible.
+This is a proposal that explains what we should do to achieve a cross-platform contiainer management system.
```
nit: `/s/contiainer/container`

ixdy commented on docs/proposals/multi-platform.md:
```diff
@@ -0,0 +1,451 @@
+
+**Author**: Lucas Käldström ([@luxas](https://github.com/luxas))
+
+**Status** (05/06/2016): Most parts are already implemented; still there's room for improvement
```
minor nit: I can't tell if this date is US or rest-of-world format. Maybe use YYYY-MM-DD to be clear?

luxas commented:
@vishh @david-mcmahon @thockin @ixdy PTAL

Also, from the docker team @tianon and @estesp might be interested

thockin commented:
This is an interesting read.  It's a mix of background info, things that are already done, and some open issues for the future.  I'd really like to see those facets split out, so it is easier to figure out how much work is left on this, what's next to do, and what external deps we have.  E.g. when Docker delivers multi-arch images, how much of what we've done will change?

luxas commented:
Thanks
Yes, the initial goal was just to write down things that are relevant to this topic, and that's quite a mix
If you have an outline in mind already, feel free to post it

The kind of problematic thing with creating a proposal like this is,
 - it isn't such a well-known thing (cross-compilation in all flavours)
 - it touches many areas
 - there are external deps (docker 1.10+ and registry 2.3+ for pulling and pushing manifests)
 - I started working on it before writing a proposal, because at that time, I didn't know what multiarch k8s would look like

So all comments appreciated.

> E.g. when Docker delivers multi-arch images, how much of what we've done will change?

Not much. We will still build `binary-arch:version` images like now, but the `binary:version` image should be a manifest list, and it should be used in yaml files and everywhere where one is referencing an image. Basically, `binary:version` is an alias for `binary-${GOARCH}:version` then.

I'm investigating if we can take advantage of @estesp's [project](https://github.com/estesp/manifest-tool) already for v1.4 for this, given the 1.10+ docker requirement (btw, is kubernetes.gcr.io a v2.3+ registry?)

cc @jfrazelle

jfrazelle commented:
Along with this we should scale the testing infrastructure across platforms, we did this for docker, and maybe we can reuse some of the logic for the ["go build dashboard"](https://build.golang.org/). Otherwise stuff will constantly be broken on the other arches

I don't see a problem in having to change things when docker supports multi-arch images. It should be a streamlined change, and +1 to @estesp tool, he is definitely one of the major people working on that effort

jfrazelle commented on docs/proposals/multi-platform.md:
```diff
@@ -0,0 +1,477 @@
+
+This is especially useful if we want to include the binary in a container. 
+If the binary is statically compiled, we may use `busybox` or even `scratch` as the base image.
+This should be the preferred way of compiling binaries that strictly require C code to be a part of it.
```
fwiw you can statically cross compile cgo from linux to all other arches, it's not something I am proud of knowing how to do though

luxas replied:
Thanks for pointing at a doc flaw.
That's exactly what my the intention is with this section, to describe how to cross-compile statically with CGO code for another architecture.

However, I see that here it says `compilation` in the header, when it should be `cross-compilation`, and likewise in other places. Will fix.

ixdy commented on docs/proposals/multi-platform.md:
```diff
@@ -0,0 +1,477 @@
+
+### Running a CI
+
+To ensure all functionality really is working on an other platform, the community should be able to setup a CI. To be able to do that, all the test-specific images have to be ported to multiple architectures. The test code also have to choose images based on the apiserver's platform.
```
In particular, Google is using the `kubekins-test` docker image on Jenkins right now, and it only runs on amd64.

https://github.com/kubernetes/test-infra/pull/257 and https://github.com/kubernetes/test-infra/pull/250 are also probably relevant here. cc @fejta

ixdy replied:
additional challenges: actually having appropriate hardware to run these tests on.

good news is that once we have CI running somewhere, it should be easy to integrate with the [test grid](http://k8s-testgrid.appspot.com/).

fejta replied:
I'm fundamentally in agreement here... but I would like to start by making our CI reproducible on one platform... then get started on other platforms.

Trying to run testing right now is pretty byzantine. So I'm hesitant to spend much time making our current testing images/scripts multi-platform when they are likely to change so dramatically in the next few months.

The linked test-infra issues are highly relevant, as well as https://github.com/kubernetes/test-infra/pull/234

fejta commented on docs/proposals/multi-platform.md:
```diff
@@ -0,0 +1,477 @@
+   - Document in the devel documentation how to do this, and make people actually care about doing it
+ - Help Docker to fully implement the Manifest List, and implement it for all Kubernetes infrastructure images.
+   - To help Docker directly might be a bit out of topic for a Kubernetes proposal, so the first step is to use/build a small utility that is able to build and push Manifest list images.
+ - Make it possible to run a CI (for distributed testing) on other platforms
```
Totally agree that we need to make it easy for anyone to setup CI, and that it should work on any platform that uses kubernetes... ideally by using kubernetes itself.

spxtr commented on docs/proposals/multi-platform.md:
````diff
@@ -0,0 +1,477 @@
+
+```console
+# From build/build-image/cross/Dockerfile when we're using Go 1.5+
+$ for platform in ${PLATFORMS}; do GOOS=${platform%/*} GOARCH=${platform##*/} go install std; done
````
If your go installation is owned by root then this will fail.

luxas commented:
I'm still working on this; soon ready for the next reviewing round.

Too much other k8s things have taken my time from this doc ;)

dims commented:
@luxas That was a great read! thanks.

k82cn commented on docs/proposals/multi-platform.md:
```diff
@@ -0,0 +1,493 @@
+
+## Use Cases
+
+TODO
```
There're the use cases from China that: both PowerLinux & X64 will be deployed in the data center, it's better for k8s to have the ability to manage those hardware in one cluster.

There're also some other requirements on how to manage heterogeneous environment, but this's the first step :).

luxas commented:
@vishh @dchen1107 @david-mcmahon @ixdy @thockin @jfrazelle @dims Here is another round of the proposal. I think we can merge this now; refactored it to make it _a lot_ clearer.

Please say what you think and apply the LGTM label if it's okay :)

luxas commented:
ping @vishh @david-mcmahon @jfrazelle PTAL

m1093782566 commented:
@luxas +1 from me

vishh commented on docs/proposals/multi-platform.md:
```diff
@@ -0,0 +1,554 @@
+### Support all platforms in a "run everywhere" deployment
+
+The easiest way of running Kubernetes on another architecture is probably by using the docker or the docker-multinode deployment. Of course, you may choose whatever deployment you want, the binaries are easily downloadable from the URL above. But [docker-multinode](https://github.com/kubernetes/kube-deploy/tree/master/docker-multinode) should be a "just works" multi-platform solution with docker  ...
+Note: That doesn't mean _everything_ "just works" with the docker-multinode setup, but basic functionality is there.
```
I'm not confident that docker multinode is an acceptable production solution. It might be worth mentioning that. It is mainly meant for "kick the tires" experience.

luxas replied:
No it isn't a production env deployment; and that's why I wrote the _basic_ functionality is there (it nearly passes all tests; I think there were 5 fails of all e2e tests)

I can clarify that's it's meant for a "kick the tires"-cluster even more if you want me to

vishh commented on docs/proposals/multi-platform.md:
```diff
@@ -0,0 +1,554 @@
+
+#### Vendored packages
+
+Kubernetes might depend on packages that aren't coded in a cross-platform manner.
```
Wouldn't build CI catch bad deps in a PR?

luxas replied:
No and yes.

They should be, but `make release` isn't run as a presubmit test right now.
Example PR were this happened: https://github.com/kubernetes/kubernetes/pull/29349

That turned out to be a really bad idea [when releasing](https://github.com/kubernetes/test-infra/issues/393#issuecomment-240320412) however, as it delayed the `v1.4.0-alpha.3` release with more than a week,

But I might add that the goal should be that the CI should directly catch such failures (although it isn't so today)

vishh replied:
Without CI, expect more such delays in the future. I'd gate multi-arch support on CI personally.

luxas commented:
@vishh Please LGTM this _now._ :)
Personally, it would mean much to get this into v1.4, and I've waited a very long time for it already.
Addressed your comments.

luxas commented:
@k8s-bot test this please github issue: #IGNORE

vishh commented on docs/proposals/multi-platform.md:
```diff
@@ -0,0 +1,556 @@
+ - Relying on platform-specific syscalls (e.g. `syscall.Dup2`)
+
+If someone tries to add a dependency that doesn't satisfy these requirements; he/she should first fix the problem in the repository that is about to be added.
+When the dependency satisfies the Kubernetes requirements for compiling on all platforms, the PR can be merged.
```
The CI should catch incompatible vendor dependencies.

vishh commented on docs/proposals/multi-platform.md:
```diff
@@ -0,0 +1,556 @@
+
+In fact, we faced this problem while this proposal was being written; in [#25243](https://github.com/kubernetes/kubernetes/pull/25243). It is quite obvious that the release blocker is of higher priority.
+
+However, before temporarily [deactivating builds](https://github.com/kubernetes/kubernetes/commit/2c9b83f291e3e506acc3c08cd10652c255f86f79), the author of the breaking PR should first try to fix the problem. If it turns out being really hard to solve, builds for the affected platform may be deactivated and a P1 issue should be made to activate them again.
```
Again, if we had CI that built and ran on all supported architectures, I will feel more comfortable.
I feel less comfortable claiming support on an arch without sufficient testing.

vishh commented:
@luxas I'd gate this proposal on requiring testing for additional `arch` before claiming support for them. I don't like multi-arch blocking progress on `amd64` if CI doesn't exist for other architectures. If you can include CI as a pre-req for multi-arch, this proposal lgtm.
Feel free to self-lgtm once you add the CI requirement.

luxas committed:
Add a multiarch proposal

luxas commented:
@vishh Thanks for the review!
I added a `Support levels` section about how/when platforms are considered "stable"
Also clarified even more that CI presubmit have to exist.

Self-applying

m1093782566 commented:
:+1:
